fix(chat): upload rework

josephmcg opened the pull request:
<!--  Thanks for sending a pull request!
If this is your first time, please read our contributor guidelines: https://github.com/Satellite-im/Core-PWA/wiki/Contributing
-->

**What this PR does** 📖
### Code related
- helpful new getters
  - isGroup - determine whether active conversation is group or DM
  - recipient - fetch active recipient data (Group or User). This will be used instead of passing recipient prop all over the place. its still being used as a prop for unrelated components, but I thought that should be handled separately
  - getFiles - fetch files for active conversation based on recipient address
- generate smaller thumbnails for file uploads, should help w/ memory
  - move this thumbnail function from filesystem to utilities
- move & rename types associated w/ chat upload
- rather than updating local state inside the components for files with a subscription, move all files to store only
- rename upload nested components to be more concise
- move file count error outside preview for better v-if rendering (save memory)
- reduce redundancy for progress tracking. add to files store value rather than use a separate list just for progress
- refactor preview css to use flex (was using inline-block 😵‍💫)
- move file upload method to chatbar since everything is in store. easier than calling a method of a ref

### Functionality
- big file uploads with streams api
- allow sending of nsfw files, but attach an nsfw boolean to it.
  - if a user has the 'block nsfw content' setting on, it will blur
  - In a future update, I will add a privacy setting to blur images from people not on your friends list (provide option to click and remove blur)
- not functionality, but make upload preview closer to figma

**Which issue(s) this PR fixes** 🔨
AP-1470, AP-1650, AP-1651, AP-1844
<!--AP-X-->

**Special notes for reviewers** 🗒️

**Additional comments** 🎤
### things to come
- aforementioned privacy blur setting
- index to track chat uploads
- improve file limit ux. rather than preventing upload, we could slice the file list and add up to the 8 file limit
- show preview immediately and load thumbnail async

molimauro commented on components/views/chat/chatbar/Chatbar.html:
```diff
@@ -1,30 +1,16 @@
-  :class="ui.replyChatbarContent.id ? 'has-reply' : ''"
->
+<div id="chatbar" ref="chatbar">
   <Enhancers @click="handleChatTextFromOutside" :recipient="recipient" />
```
We can remove recipient also here right?

molimauro commented on components/views/files/upload/preview/Preview.html:
```diff
@@ -0,0 +1,18 @@
+  @touchstart="handleTouchPreview"
+  @touchend="handleTouchPreview"
+>
+  <UiScroll horizontalScroll scrollbarVisibility="scroll">
```
Can we use here `UiSimpleScroll`? We're ditching the old one thanks :)

molimauro commented on components/views/files/upload/Error.vue:
```diff
@@ -0,0 +1,39 @@
+  padding: @normal-spacing;
+  width: calc(@full - 2rem);
+  margin: 0 @normal-spacing;
+  background-color: #181f2c;
+  border-bottom-color: rgba(87, 101, 116, 0.3);
```
new colors?

josephmcg replied:
leftover from copy paste, thanks

phillsatellite commented:
One thing I found when testing, if I send an NSFW image in chat, then block NSFW when I open the image in chat should it be blurred? 
https://user-images.githubusercontent.com/93608357/173372783-c3f79afc-5346-4652-b95d-551fd89f366d.mov

phillsatellite commented:
One thing I found when testing, if I send an NSFW image in chat, then block NSFW when I open the image in chat should it be blurred? 
https://user-images.githubusercontent.com/93608357/173372783-c3f79afc-5346-4652-b95d-551fd89f366d.mov

molimauro commented on components/views/files/upload/Upload.vue:
```diff
@@ -65,192 +53,84 @@ export default Vue.extend({
 
-        if (newFiles.length + this.files.length > 8) {
-          this.$store.commit('chat/setCountError', true)
+      if (newFiles.length + this.getFiles.length > 8) {
```
8 sounds like a config ;)

molimauro commented on utilities/Thumbnail.ts:
```diff
@@ -0,0 +1,71 @@
+    if (await _tooLarge(fileJpg)) {
+      return
+    }
+    return blobToBase64(await skaler(fileJpg, { width }))
```
why we return the base64 instead of the image?

josephmcg replied:
originally this was only working within the context of file system. in that case I used base64 so I can easily add it to the json index. In this case, it doesnt need base64. I can rework that

molimauro replied:
ok perfect thanks a lot!

josephmcg commented:
@phillsatellite @molimauro 
Thank you both for taking a look. everything has been addressed

molimauro commented:
Good to me, thanks Joe!!

phillsatellite commented:
Blur when preview is fixed thanks Joe! I found 2 other things when I retested, it seems like the "view original" has moved from the left side of the preview to the center 
<img width="1728" alt="Screen Shot 2022-06-14 at 11 15 38 AM" src="https://user-images.githubusercontent.com/93608357/173613841-e7ff16db-4498-4bf8-84aa-cf704d11d347.png">

I also saw this error if I cancel a 2GB file upload in chat half way through. This error will appear each time you navigate into the chat as well
https://user-images.githubusercontent.com/93608357/173612903-a652fbd1-fee9-40ed-a8c3-47d8b1fa5a36.mov

josephmcg commented:
@molimauro I realized why we were using base64. I was encountering bugs where the rendered preview items would lose track of the images if you deleted the ones in the middle. I switched back to base64 for now since it's simpler. maybe we can revisit later. I'm glad we split up the thumbnail and base64 utils though.

@phillsatellite I adjusted the style of that 'view original' link. I removed the box shadow and improved the context due to discussions we were having in [this ticket](https://satellite-im.atlassian.net/browse/AP-1561). 
For now, I changed it so you cannot remove a file from the list once upload has started. I would like to look into cancelling uploads at a later date

josephmcg commented:
heads up that gifs will not move inside the preview. I would like to improve that separately

molimauro commented:
> @molimauro I realized why we were using base64. I was encountering bugs where the rendered preview items would lose track of the images if you deleted the ones in the middle. I switched back to base64 for now since it's simpler. maybe we can revisit later. I'm glad we split up the thumbnail and base64 utils though.
> 
> @phillsatellite I adjusted the style of that 'view original' link. I removed the box shadow and improved the context due to discussions we were having in [this ticket](https://satellite-im.atlassian.net/browse/AP-1561). For now, I changed it so you cannot remove a file from the list once upload has started. I would like to look into cancelling uploads at a later date

Ok perfect @josephmcg, thanks for explaining! All good then ;)

stavares843 reviewed (changes requested):
dev

<img width="685" alt="Captura de ecrã 2022-06-15, às 23 47 14" src="https://user-images.githubusercontent.com/29093946/173954948-5e431da8-1f15-464e-9ee2-0fcd5e248469.png">


this branch

<img width="797" alt="Captura de ecrã 2022-06-15, às 23 48 34" src="https://user-images.githubusercontent.com/29093946/173954969-2ca38fe5-0164-48e7-8b4d-458bb13cbfd1.png">


same image, same account, same browser

josephmcg commented:
this did not occur for me. could you make sure you have the latest changes and you `yarn dev` after switching to this branch? not switching while it is already running
![image](https://user-images.githubusercontent.com/33670767/173969884-e99541a6-5acf-4703-9a46-1c502af0a48c.png)

stavares843 commented:
no longer happens

strange, i also did that yesterday and was still reproducible


also

dev vs this branch

<img width="809" alt="Captura de ecrã 2022-06-16, às 15 59 14" src="https://user-images.githubusercontent.com/29093946/174099630-a5acf07a-6ce8-4f3c-b425-198afb3f0bda.png">
<img width="756" alt="Captura de ecrã 2022-06-16, às 15 59 19" src="https://user-images.githubusercontent.com/29093946/174099653-b0508078-6822-45f6-ac88-adefcd6fdd37.png">


on dev, if you send an image and fails to load such as an SVG, appears the failed to load placeholder, but on this branch doesn't appear

josephmcg commented:
I do not have that issue :/
I just rebased, hoping that will fix it for you

![image](https://user-images.githubusercontent.com/33670767/174236076-649572e4-b5bd-498d-a8e3-e48c5ddf8302.png)
